katdal/sensordata.py

```diff
@@ -126,34 +190,96 @@ class TelstateSensorData(SensorData):
     Parameters
     ----------
     data : recarray-like, with fields ('timestamp', 'value')
-        Uncached sensor data as record array or equivalent (such as a
-        :class:`h5py.Dataset`)
+        Uninterpolated sensor data as structured array or equivalent (such as
+        an :class:`h5py.Dataset`)
     name : string or None, optional
         Sensor name (assumed to be data.name by default, if it exists)
 
-    Attributes
+    """
+
+    def __init__(self, data, name=None):
+        super(H5TelstateSensorData, self).__init__(data, name)
+        # The dtype is not immediately available - need to unpickle data first
+        self.dtype = None
+
+    def __getitem__(self, key):
+        """Extract timestamp and value of each sensor data point."""
+        if key == 'timestamp':
+            return np.asarray(self._data[key])
+        elif key == 'value':
+            # Unpack everything first, otherwise old files will be a mess
+            values = [_h5_telstate_unpack(s) for s in self._data[key]]
+            # Figure out dtype and wrap any objects
+            self.dtype = infer_dtype(values)
+            if self.dtype == np.object:
+                values = [ComparableArrayWrapper(value) for value in values]
+            return values
+        else:
+            raise ValueError("Sensor %r data has no key '%s'" % (self.name, key))
+
+
+class TelstateSensorData(SensorData):
+    """Raw (uninterpolated) sensor data stored in original TelescopeState.
+
+    This wraps sensor data stored in a TelescopeState object. The data is
+    only read out on item access and then cached on the object. The caching
```

Comment:
Maybe explain the motivation for caching here i.e. to ensure that requesting timestamps and then values only loads the data once instead of twice. At first I was confused why this classes does caching while H5TelstateSensorData does not.